pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,394 @@
+					e.Update(e.relistDuration) // Update the relisting period to the default value for the Generic PLEG.
+					e.genericPleg.Start()
+					break
+				}
+			}
+
+			err := e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+			if err != nil {
+				numAttempts++
+				e.Relist() // Force a relist to get the latest container and pods running metric.
+				klog.V(4).InfoS("Evented PLEG: Failed to get container events, retrying: ", "err", err)
+			}
+		}
+	}()
+
+	if isEventedPLEGInUse() {
+		e.processCRIEvents(containerEventsResponseCh)
+	}
+}
+
+func (e *EventedPLEG) processCRIEvents(containerEventsResponseCh chan *runtimeapi.ContainerEventResponse) {
+	for event := range containerEventsResponseCh {
+		podID := types.UID(event.PodSandboxStatus.Metadata.Uid)
+		shouldSendPLEGEvent := false
+
+		status, err := e.runtime.GeneratePodStatus(event)
+		if err != nil {
+			// nolint:logcheck // Not using the result of klog.V inside the
+			// if branch is okay, we just use it to determine whether the
+			// additional "podStatus" key and its value should be added.
+			if klog.V(6).Enabled() {
+				klog.ErrorS(err, "Evented PLEG: error generating pod status from the received event", "podUid", podID, "podStatus", status)
+			} else {
+				klog.ErrorS(err, "Evented PLEG: error generating pod status from the received event", "podUid", podID, "podStatus", status)
+			}
+		} else {
+			if klogV := klog.V(6); klogV.Enabled() {
+				klogV.InfoS("Evented PLEG: Generated pod status from the received event", "podUid", podID, "podStatus", status)
+			} else {
+				klog.V(4).InfoS("Evented PLEG: Generated pod status from the received event", "podUid", podID)
+			}
+			// Preserve the pod IP across cache updates if the new IP is empty.
+			// When a pod is torn down, kubelet may race with PLEG and retrieve
+			// a pod status after network teardown, but the kubernetes API expects
+			// the completed pod's IP to be available after the pod is dead.
+			status.IPs = e.getPodIPs(podID, status)
+		}
+
+		e.updateRunningPodMetric(status)
+		e.updateRunningContainerMetric(status)
+
+		if event.ContainerEventType == runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT {
+			for _, sandbox := range status.SandboxStatuses {
+				if sandbox.Id == event.ContainerId {
+					e.cache.Delete(podID)
+				}
+			}
+			shouldSendPLEGEvent = true
+		} else {
+			if e.cache.Set(podID, status, err, time.Unix(event.GetCreatedAt(), 0)) {
+				shouldSendPLEGEvent = true
+			}
+		}
+
+		if shouldSendPLEGEvent {
+			e.processCRIEvent(event)
+		}
+	}
+}
+
+func (e *EventedPLEG) processCRIEvent(event *runtimeapi.ContainerEventResponse) {
+	switch event.ContainerEventType {
+	case runtimeapi.ContainerEventType_CONTAINER_STOPPED_EVENT:
+		e.sendPodLifecycleEvent(&PodLifecycleEvent{ID: types.UID(event.PodSandboxStatus.Metadata.Uid), Type: ContainerDied, Data: event.ContainerId})
+		klog.V(4).InfoS("Received Container Stopped Event", "event", event)
```

Comment:
not 100% sure if it's needed here and below, but do you need to call `event.String()` to ensure you get string representation of the event in the log?

Reply:
Yeah, used `event.String()` now.